src/flow-worker.c

```diff
@@ -380,6 +380,36 @@ static inline void FlowWorkerStreamTCPUpdate(ThreadVars *tv, FlowWorkerThreadDat
     /* Packets here can safely access p->flow as it's locked */
     SCLogDebug("packet %"PRIu64": extra packets %u", p->pcap_cnt, fw->pq.len);
     Packet *x;
+    while ((x = PacketDequeueNoLock(&tv->decode_pq))) {
+        // switch flow to new protocol
+        void *alstate_orig = p->flow->alstate;
+        AppProto alproto_orig = p->flow->alproto;
+        // hardcoded logic
+        if (alproto_orig == ALPROTO_HTTP2) {
+            p->flow->alproto = ALPROTO_DNS;
+            p->flow->alstate = SCHttp2GetLayeredState(p->flow->alstate);
+        }
+        // run parsing, detection and logging
+        AppLayerParserParse(tv, NULL, p->flow, p->flow->alproto,
+                (x->flowflags & FLOW_PKT_TOSERVER) ? STREAM_TOSERVER : STREAM_TOCLIENT, x->payload,
+                x->payload_len);
+        if (detect_thread != NULL) {
+            FLOWWORKER_PROFILING_START(x, PROFILE_FLOWWORKER_DETECT);
+            Detect(tv, x, detect_thread);
+            FLOWWORKER_PROFILING_END(x, PROFILE_FLOWWORKER_DETECT);
+        }
+        OutputLoggerLog(tv, x, fw->output_thread);
+        FramesPrune(x->flow, x);
+
+        // switch back to real protocol
+        p->flow->alstate = alstate_orig;
+        p->flow->alproto = alproto_orig;
+
+        // releases memory associated to the packet
+        SCFree(x->payload);
+        FlowDeReference(&x->flow);
+        TmqhOutputPacketpool(tv, x);
+    }
```

Comment:
⚠️ this is the third key step

We deque and process the packets, switching the protocol and state...